src/peer_loop.rs

```diff
@@ -203,6 +206,87 @@ impl PeerLoopHandler {
         sanction_result.map_err(|err| anyhow::anyhow!("Cannot reward banned peer: {err}"))
     }
 
+    /// Update a peer's bootstrap status.
+    ///
+    /// Per connection, the bootstrap status can be set once without
+    /// incurring [punishment](Self::punish). After a cooldown period, the
+    /// bootstrap status can be set again. This is to prevent the peer from
+    /// forcing a write-lock acquisition on the global state lock.
+    ///
+    /// This method only acquires a write lock for the global state lock if
+    /// necessary.
+    ///
+    /// # Locking:
+    ///   * acquires `global_state_lock` for read
+    ///   * might acquire `global_state_lock` for write
+    async fn set_bootstrap_status(&mut self, status: BootstrapStatus) -> Result<()> {
```

Comment:
As with the issue in `main_loop`, this method takes locks a bit too far into the callstack for my taste. However, what this method actually is, is the handling of `PeerMessage::Bootstrap`, so if we were to factor out the handling of all peer messages (please don't do that in this PR) to separate methods, this pattern would make more sense, I think. I think just renaming this method to "handle_boostrap_message" or something similar would fix my concern -- that might even provide a nice template for a later rewrite of `handle_peer_message`.